lib/common/rules.c

```diff
@@ -595,6 +596,23 @@ pcmk__evaluate_date_expression(const xmlNode *date_expression,
     return rc;
 }
 
+/*!
+ * \internal
+ * \brief Check whether a given regular expression submatch is available
+ *
+ * \param[in] submatches  Regular expression submatches (as set by regexec())
+ * \param[in] nmatches    Number of entries in \p submatches[]
+ * \param[in] i           Index of submatch to check for
+ *
+ * \return true if \p submatches has submatch \p i, otherwise false
+ */
+static inline bool
+valid_submatch(const regmatch_t submatches[], int nmatches, int i)
+{
+    return (nmatches > i) && (submatches[i].rm_so != -1)
+           && (submatches[i].rm_eo > submatches[i].rm_so);
```

Comment:
Can `rm_eo <= rm_so` happen, and should we really treat it as invalid if so? The `regmatch_t` man page doesn't actually explain `rm_so` and `rm_eo` but it looks like they're start offset and end offset, respectively, and that this would indicate an empty submatch (length 0).

I can't say I understand much of `pcmk__replace_submatches()` yet, to be fair. Trusting that it works as it should, until it's time for a best practices commit or something.

Reply:
An empty match is possible (like `a(.*)b` matching `ab`) but rm_so is the index of the first matching character, so I was thinking rm_so == -1 for empty matches same as no match. However I tested it and rm_so == rm_eo == index after the pattern. I think it would make sense to strip `%N` for empty matches.

Reply:
That would be a change in behavior though, so I'll save it for 3.0

Reply:
Ah, the code already strips `%N` for both empty matches and non-matches, so it's not an issue. I will add test cases for those and some more comments.

Reply:
Memory will currently be overallocated by 2 bytes per empty match or non-match. I think I can refactor it to be easier to follow and avoid that.